services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java

```diff
@@ -227,53 +236,77 @@ protected UserProfileMetadata decorateUserProfileForCache(UserProfileMetadata me
             if (rc != null && !(UserModel.USERNAME.equals(attributeName) || UserModel.EMAIL.equals(attributeName))) {
                 // do not take requirements from config for username and email as they are
                 // driven by business logic from parent!
-
-                if (rc.isAlways() || UPConfigUtils.isRoleForContext(context, rc.getRoles())) {
+                required = AttributeMetadata.ALWAYS_TRUE;
+                if (rc.isAlways()) {
                     validators.add(createRequiredValidator(attrConfig));
-                    required = AttributeMetadata.ALWAYS_TRUE;
+                } else if (rc.getRoles() != null) {
+                    validators.add(new AttributeValidatorMetadata(AttributeRequiredByMetadataValidator.ID,
+                            ValidatorConfig.builder()
+                                    .config("roles", rc.getRoles())
+                                    .build()));
                 } else if (UPConfigUtils.canBeAuthFlowContext(context) && rc.getScopes() != null && !rc.getScopes().isEmpty()) {
                     // for contexts executed from auth flow and with configured scopes requirement
                     // we have to create required validation with scopes based selector
-                    required = (c) -> attributePredicateAuthFlowRequestedScope(rc.getScopes());
-                    validators.add(createRequiredValidator(attrConfig));
+                    validators.add(new AttributeValidatorMetadata(AttributeRequiredByMetadataValidator.ID,
```

Comment:
`required` predicate is not set correctly here, so metadata used in forms are incorrect, they always mark attribute as required